server/src/controllers/attainable.ts

```diff
@@ -0,0 +1,255 @@
+  const date: Date = req.body.date;
+  const expiryDate: Date = req.body.expiryDate;
+  const requestSubAttainables: Array<AttainableRequestData> | undefined = req.body.subAttainments;
+
+  // If linked to a parent id, check that it exists and belongs to the same course instance.
+  if (parentId) {
+    const parentAttainable: Attainable = await findAttainableById(
+      parentId, HttpCode.UnprocessableEntity
+    );
+
+    if (parentAttainable.courseInstanceId !== courseInstanceId) {
+      throw new ApiError(
+        `parent attainment ID ${parentId} does not belong ` +
+        `to the course instance ID ${courseInstanceId}`,
+        HttpCode.Conflict
+      );
+    }
+  }
+
+  const attainable: Attainable = await models.Attainable.create({
+    courseId: courseId,
+    attainableId: parentId,
+    courseInstanceId: courseInstanceId,
+    name: name,
+    date: date,
+    expiryDate: expiryDate
+  });
+
+  if (!requestSubAttainables  || requestSubAttainables.length === 0) {
+    res.status(HttpCode.Ok).json({
+      success: true,
+      data: {
+        attainment: {
+          id: attainable.id,
+          courseId: attainable.courseId,
+          courseInstanceId: attainable.courseInstanceId,
+          name: attainable.name,
+          date: attainable.date,
+          expiryDate: attainable.expiryDate,
+          parentId: attainable.attainableId,
+          tag: generateAttainableTag(
+            attainable.id, attainable.courseId, attainable.courseInstanceId
+          )
+        }
+      }
+    });
+    return;
+  }
+
+  async function processSubAttainables(
+    unprocessedAttainables: Array<AttainableRequestData>, parentId: number
+  ): Promise<Array<AttainableData>> {
+    const attainables: Array<AttainableData> = [];
+    let subAttainables: Array<AttainableData> = [];
+
+    for (const attainable of unprocessedAttainables) {
+      const dbEntry: Attainable = await models.Attainable.create({
+        attainableId: parentId,
+        courseId: courseId,
+        courseInstanceId: courseInstanceId,
+        name: attainable.name,
+        date: attainable.date,
+        expiryDate: attainable.expiryDate
+      });
+
+      if (attainable.subAttainments.length > 0) {
+        subAttainables = await processSubAttainables(attainable.subAttainments, dbEntry.id);
+      }
+
+      attainables.push({
+        id: dbEntry.id,
+        courseId: dbEntry.courseId,
+        courseInstanceId: dbEntry.courseInstanceId,
+        name: dbEntry.name,
+        date: dbEntry.date,
+        expiryDate: dbEntry.expiryDate,
+        parentId: dbEntry.attainableId,
+        tag: generateAttainableTag(
+          dbEntry.id, dbEntry.courseId, dbEntry.courseInstanceId
+        ),
+        subAttainments: subAttainables
+      });
+    }
+    return attainables;
+  }
+
+  const subAttainables: Array<AttainableData> = await processSubAttainables(
+    requestSubAttainables, attainable.id
+  );
+
+  res.status(HttpCode.Ok).json({
```

Comment:
I think it would be better not to have two `res.json` calls depending on whether sub attainables exist. Instead there should be a single `res.json` call at the end, with an `AttainableData` object which is modified depending on whether sub attainables exist. Pseudo-code below to illustrate what I mean.

Currently:
```ts
if (!subAttainables) {
  res.send(/* attainables */);
  return;
}

processSubAttainables(...);
res.send(/* attainables + subAttainables */);
```
Instead:
```ts
attainablesData: AttainableData = { ... };

if (subAttainables) {
  attainablesData.subAttainments = processSubAttainables(...);
}

res.send(attainablesData);
```
It may also be better to return an empty array in `attainablesData.subAttainments` if no sub attainments exist rather than leaving it undefined.

Reply:
Good point, fixed.